community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -35,33 +43,21 @@ case class ExpandIntoPipe(source: Pipe,
                           types: LazyTypes)(val estimatedCardinality: Option[Double] = None)
                          (implicit pipeMonitor: PipeMonitor)
   extends PipeWithSource(source, pipeMonitor) with RonjaPipe {
-
   self =>
 
+  private val CACHE_SIZE = 100000
+  //cache of known connected nodes
+  private val connected = new LRUCache[(Node, Node), Seq[Relationship]](CACHE_SIZE)
```

Comment:
Right it could probably be cached in the execution-context or query-state. As we're eager, i.e. match using the information before this cypher query updated the graph?